core/src/main/java/com/google/cloud/sql/core/AsyncRateLimiter.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A simple, constant-time rate limit calculator. Ensures that there is always at least
+ * delayBetweenAttempts milliseconds between attempts.
+ */
+class AsyncRateLimiter {
+  private long nextOperationTimestamp;
+  private final long delayBetweenAttempts;
+
+  AsyncRateLimiter(long delayBetweenAttempts) {
+    this.delayBetweenAttempts = delayBetweenAttempts;
+  }
+
+  /**
+   * Returns the number of milliseconds to delay before proceeding with the rate limited operation.
+   * If this returns > 0, the operation must call "acquire" again until it returns 0.
+   */
+  @VisibleForTesting
+  synchronized long nextDelayMs(long nowTimestampMs) {
+    // allow exactly 1 operation to pass the timestamp.
+    if (nextOperationTimestamp <= nowTimestampMs) {
+      nextOperationTimestamp = nowTimestampMs + delayBetweenAttempts;
+      return 0;
+    }
+
+    return nextOperationTimestamp - nowTimestampMs;
+  }
+
+  /**
+   * Returns a future that will be done when the rate limit has been acquired.
+   *
+   * @param executor the executor to use to schedule future checks for available rate limits.
+   */
+  public ListenableFuture<Long> acquireAsync(ScheduledExecutorService executor) {
```

Comment:
I've switched this to `ListenableFuture<Void>`. Our linter throws a warning when we use `ListenableFuture` because it doesn't allow us to use a parameterized class without type parameters.

Reply:
Actually, I had to switch back to making an internal class to track when the rate limit was acquired. Now it's called `RateLimitAcquisition` and it is package-scope and labeled with `@VisibileForTesting`. 

Unfortunately, there is way too much variability in the thread executors across all the platforms to test this with the public api. Measuring the time that acquireAsync() resolved using `Future.addListener()` in the test case introduced much delay on Windows and sometimes Mac VMs. This make the tests flakey.